### ISSUE TITLE
noise() is not Perlin noise

### ISSUE THREAD

**Issue by cheind** [processing/p5.js#7430]
### Topic

Hi,

I've been browsing the documentation of the noise function 

> [...] Ken Perlin invented `noise()` while animating the original Tron film in the 1980s

and continued to study its implementation. From a first sight, I believe that the implementation found in p5.js deviates from Perlin noise in two characteristics:

### Perlin noise defines gradients at integer grid locations

The noise value associated with a grid locations is given by the dot-product between the stored gradient and the offset vector. If I read your implementation correctly, you are directly assigning a random value to each integer location that is later interpolated.

### Perlin noise is zero at all integer grid locations
if the dot-product becomes zero, either because of orthogonality between the gradient direction and the offset vector or because the offset vector is zero (at integer locations), the resulting noise value becomes zero. Hence, at integer locations the resulting noise should be zero. In the original implementation the noise is bounded between [-sqrt(2)/2, +sqrt(2)/2] for the 2D case. In the below plot, the iso-contour of noise==0 is shown as green line.

![image](https://github.com/user-attachments/assets/6d071fbe-315c-4f84-ad60-fab495a22503)
(taken from https://en.wikipedia.org/wiki/Perlin_noise)

in p5.js  `noise(x=0,0,0)` will not return a zero noise value in general.

Hence, `noise()` provides smooth noise, but not Perlin noise. I don't believe that's an issue for the intended audience, but in case one relies on the above Perlin characteristics to hold true, you should mentioned the deviations in the docs.

**Comment by subhraneel2005** [comment]
Hi, thank you pointing out the differences between the current implementation and traditional Perlin noise.

I tried to fix it, and this is what i did, please review them:
- modified the function to define gradients at integer grid locations, as Perlin originally intended.
- ensured that the noise value is zero at integer points that aligning with the expected behavior

My Solution 👇
```
p5.prototype.noise = function(x, y = 0, z = 0) {  
  if (perlin == null) {  
    perlin = new Array(PERLIN_SIZE + 1);  
    for (let i = 0; i < PERLIN_SIZE + 1; i++) {  
      // Assigned random gradient vectors instead of random values  
      let angle = Math.random() * Math.PI * 2;  
      perlin[i] = { x: Math.cos(angle), y: Math.sin(angle) }; // 2D gradient  
    }  
  }  

  if (x < 0) x = -x;  
  if (y < 0) y = -y;  
  if (z < 0) z = -z;  

  let xi = Math.floor(x),  
      yi = Math.floor(y),  
      zi = Math.floor(z);  
  let xf = x - xi;  
  let yf = y - yi;  
  let zf = z - zi;  

  let r = 0;  
  let ampl = 1.0; // Started with full amplitude  

  for (let o = 0; o < perlin_octaves; o++) {  
    let of = xi + (yi << PERLIN_YWRAPB) + (zi << PERLIN_ZWRAPB);  

    // Calculated dot product with gradients  
    let n1 = dotProduct(perlin[of & PERLIN_SIZE], xf, yf);  
    let n2 = dotProduct(perlin[(of + PERLIN_YWRAP) & PERLIN_SIZE], xf, yf);  
    n1 += scaled_cosine(yf) * (n2 - n1);  

    let n3 = dotProduct(perlin[(of + PERLIN_ZWRAP) & PERLIN_SIZE], xf, yf);  
    let n4 = dotProduct(perlin[(of + PERLIN_YWRAP + PERLIN_ZWRAP) & PERLIN_SIZE], xf, yf);  
    n3 += scaled_cosine(yf) * (n4 - n3);  

    n1 += scaled_cosine(zf) * (n3 - n1);  

    r += n1 * ampl;  
    ampl *= perlin_amp_falloff;  
    
    // Updated xi, yi, zi for the next octave  
    xi <<= 1;  
    xf *= 2;  
    yi <<= 1;  
    yf *= 2;  
    zi <<= 1;  
    zf *= 2;  

    // Wraped around logic  
    if (xf >= 1.0) {  
      xi++;  
      xf--;  
    }  
    if (yf >= 1.0) {  
      yi++;  
      yf--;  
    }  
    if (zf >= 1.0) {  
      zi++;  
      zf--;  
    }  
  }  

  return r;  
};
```

**Comment by cheind** [comment]
@subhraneel2005 that was quick :) I haven't reviewed your code (yet). I wanted to emphasis that transitioning to Perlin noise might have unintended side-effects for users of p5.js. The zero-noise-at-grid-locations might be an issue if you populate a noise image using sampling positions that fall on integer grid locations, i.e. pixels -> zero image :)

**Comment by limzykenneth** [comment]
The implementation in p5.js comes directly from Processing and I believe the intention is for a more useful noise function that is random based. It is fine to edit the documentation to indicate this minor difference but I don't think we will be changing the implementation.

**Comment by subhraneel2005** [comment]
> The implementation in p5.js comes directly from Processing and I believe the intention is for a more useful noise function that is random based. It is fine to edit the documentation to indicate this minor difference but I don't think we will be changing the implementation.

alright then, i will not change any code implementaion. thanks for letting me know :)

**Comment by davepagurek** [comment]
It's true that the p5 noise implementation is not quite Perlin noise, and is based on the Processing noise implementation (which [I hear is based on demoscene code from 2001](https://forum.processing.org/one/topic/recreating-processing-toxiclibs-noise-function-in-glsl-shader.html), possibly for code size or performance constraints that were more important when it was first added?) 

Anyway, this topic has definitely come up before, but there's balance we have to find with how complex we keep the reference. Especially in explaining what the deviation is -- talking about dot products with vector offsets seems maybe a little too much technical detail for the p5.js reference. One option might be to use actual Perlin noise, although it has been brought up that, at this point, being able to use the same noise function that has been in Processing from the start is also kinda important. We've also thought about having different noise modes, where a library could possibly implement e.g. simplex noise.

@limzykenneth @ksen0 let me know if you have any thoughts on this one!

**Comment by davepagurek** [comment]
oh haha I see you added a comment in the mean time while Github was having issues

**Comment by cheind** [comment]
@davepagurek Agreed. I think that changing the implementation is an overkill here - but I believe the deviations should be mentioned in case anyone is using p5.js for scientific visualizations. Probably one can shorten this to something along the lines:

> The noise() function generates smooth noise similar to Perlin noise but differs in two key ways: it is not gradient-based, and its zero-noise iso-contours do not fall on integer grid-locations in general.

**Comment by cheind** [comment]
@davepagurek you may need to sift through examples as well. Its called Perlin noise at quite a few locations.

**Comment by cheind** [comment]
@SableRaf , I believe `noise()` is a value noise generator, since the generated noise stems from random values assigned to grid locations (whereas gradient noise would assign random gradients to grid locations). From my understanding, simplex noise is a variant of gradient noise, just with better runtime properties in higher dims (naive impl requires O(2**dims) dots/interpolations, and simplex noise reduces this to O(dims**2)).

As previously mentioned, I agree that re-writing the implementation solely to align with the documentation would be misguided. Gradient noise possesses unique properties that might surprise your user base. However, it also offers distinct advantages over value noise. Notably, gradient noise places greater emphasis on higher frequencies compared to lower frequencies, which helps to reduce the prevalence of flat areas in the noise landscape. Side note, the hash function used in the current implementation may additionally limit the randomness of the output (see #7431).

As @davepagurek suggested, providing a link to this discussion or a page with detailed clarifications would be a good starting point for readers seeking more information.

**Comment by SableRaf** [comment]
For added context, this has been discussed within Processing in the past: [Perlin noise documentation #51](https://github.com/processing/processing-docs/issues/51)

A `noiseMode()` function was suggested, but that was ruled out, likely to keep the core libraries simple.

**Comment by nickmcintyre** [comment]
Good catch @cheind! Yeah, I probably went too far with connecting the reference to an important bit of history -- agreed that I should have just corrected the inaccuracy.

@SableRaf's suggested revision strikes a nice balance. I believe there's one minor typo:

> ...called "[value noise](https://en.wikipedia.org/wiki/Value_noise#:~:text=Value%20noise%20is%20a%20type,Perlin%20noise%20and%20Simplex%20noise)," which...

**Comment by cheind** [comment]
@nickmcintyre these things happen all the time :) Right now, I'm sifting through the 1985 [original paper](https://dl.acm.org/doi/pdf/10.1145/325165.325247) to double-check and indeed he introduces a random value  `d` alongside the gradient at each lattice location. He mentions that this value is returned for integer locations. Its not clear to me what role `d` plays for non-lattice locations. However, he also metions that quote

> The author has developed a number of surprisingly different
implementations of the Noise() function. Some real tradeoffs are
involved between time, storage space, algorithmic complexity,
and adherence to the three defining statistical constraints.
Because of space limitations, we will describe only the simplest
such technique. 

So who knows what's truly Perlin noise after all :)

**Comment by davepagurek** [comment]
The real Perlin noise is the friends you make along the way ;)

**Comment by postspectacular** [comment]
Just to give a little more historic perspective here:

I ported that Farbrausch code for processing.core back in summer 2003. Back then there:

1) wasn't really that much easy-to-find literature or readily available sample code, though two of the most cited & helpful refs was this amazing [Hugo Elias article](https://web.archive.org/web/20030801211535/https://freespace.virgin.net/hugo.elias/models/m_perlin.htm) and the noise treatment by [the god father of computer graphics](http://www.paulbourke.net/fractals/noise/). 
2) Computers were much slower (and Java too), so for something like `noise()` to be generally useful (esp. the 2D and 3D versions), we wanted (needed!) to choose a performant implementation and so ended up with the Farbrausch version, which provided the best compromise (also was much faster than other impls I tried & ported back then)...
3) I think it's fair to say that neither Ben, Casey nor I back then considered any of the finer nuances/differences between noise types too much (or at all 😉 )... I proposed a noise() function because I thought it'd be super useful for a lot of different applications and I'd already used it for some terrain & texture generation tasks in a few games I'd developed in Shockwave3D/Director and got some very interesting results from it...
4) I'd have never thought it would stay around (largely unchanged) for that long...

Sorry for any inconvenience caused, heh! :)

**Comment by SableRaf** [comment]
Thanks @postspectacular for the historical context! Grateful you took the time to share 😃

**Comment by SableRaf** [comment]
I talked to Casey about this earlier today. He didn't remember as much as Karsten, but he had a good suggestion regarding the documentation. Do mention at some point that there are different kinds of `noise()` functions—including Perlin, Simplex, etc.—and note that the specific implementation here is value noise, but overall focus more on what `noise()` actually _is_ in relationship to a `random()` function, and less on any specific implementation. I like that.

**Comment by SableRaf** [comment]
While we're going down rabbit holes, here's another one. The claim on the p5.js documentation that noise was "_invented by Ken Perlin while animating the original Tron film in the 1980s_" seemed suspicious, especially since the [Wikipedia page for Perlin noise](https://en.wikipedia.org/wiki/Perlin_noise) (likely source for that factoid) says Perlin invented `noise()` "_after_ working on Disney's computer animated sci-fi motion picture Tron (1982)" (emphasis mine). 

Besides, the 3D models in TRON don't have any textures 🤔

So I had a look at the source for that reference: [In the beginning: The Pixel Stream Editor](https://userpages.cs.umbc.edu/olano/s2002c36/ch04.pdf), in Chapter 4 of this [SIGGRAPH 2002 Course Notes on Real-Time Shading Languages](https://userpages.cs.umbc.edu/olano/s2002c36/). 

Here is what Ken Perlin writes (with added ellipses for concision):

> Working on TRON was a blast, but on some level I was frustrated by the fact that everything looked machine-like (a typical scene is shown below). In fact, that machinelike aesthetic became the "look" associated with MAGI in the wake of TRON. So I got to work trying to help us break out of the "machine-look ghetto." (...)

<img width="400" alt="Three geometric motorcycles speeding in a corridor. Still frame of the light cycle racing scene from the movie TRON (1982)" src="https://github.com/user-attachments/assets/f6e89e44-2e14-41a8-b0b9-2370ddc13388" />

> Unfortunately (or fortunately, depending on how you look at it) our Perkin-Elmer and Gould SEL computers, while extremely fast for the time, had very little RAM, so we couldn't fit detailed texture images into memory. I started to look for other approaches. (...)

> The first thing I did in 1983 was to create a primitive space-filling signal that would give an impression of randomness. (...) I ended up developing a simple pseudo-random "noise" function that fills all of three dimensional space. A slice out of this stuff is pictured.

<img width="400" alt="A circular slice of Perlin noise on a black background" src="https://github.com/user-attachments/assets/1a73c142-4776-4043-9f8c-ce7ef2ea58dd" />

How cool is that?

In summary, Perlin started to think seriously about procedural textures while working on Tron in 1981, precisely because they could not use regular textures. Only later—in 1983—did he develop the `noise()` function to address limitations of texturing and rendering systems at the time. 

So in conclusion, it seems very probable that Ken Perlin did NOT invent `noise()` "while" working on TRON, but "after", as is correctly stated on the Wikipedia page for Perlin noise.

Thank you for ~~watching my 2h long YouTube video essay~~ reading my comment.

PS: On a side note, it's funny to see that Perlin noise is still being used in the exact way it was intended:

> Noise itself doesn't do much except make a simple pseudo-random pattern. But it provides seasoning to help you make things irregular enough so that you can make them look more interesting.

**Comment by cheind** [comment]
For those interested, I reached out to Ken for historical clarification. In particular on the scalar `d` introduced in his original paper and mentioned in my reply above : https://github.com/processing/p5.js/issues/7430#issuecomment-2551203792

I keep everyone posted, once I receive a reply (or maybe he finds the time to post here).

**Comment by cheind** [comment]
Hey,

as noted in https://github.com/processing/p5.js/issues/7430#issuecomment-2556820919, I kindly asked Ken Perlin to clarify the following point

> [...] In studying your paper, An Image Synthesizer [2], I saw that you reference a scalar value `d` alongside the gradient at each lattice point and note that `d` is returned at integer lattice coordinates. This seems to suggest a different behavior than the general description on Wikipedia [3], which defines Perlin noise as a scalar-valued multivariate function with roots at integer lattice points. At first glance, this seems contradictory.

to which Ken replied (quoting with permission)

> The goal of noise is to produce a pseudo-random signal that minimizes the spectral energy at scales different from the scale of the unit lattice. In practice I always set d = 0, because this minimizes the spectral energy at other scales.

Wishing everyone a smooth start into the new year.

**Comment by Zearin** [comment]
> > The `noise()` function is similar to "Perlin noise," a popular technique for for generating smooth, random-like patterns, invented by Ken Perlin while animating the original _Tron_ film in the 1980s. Note that `noise()` is not an exact implementation of Perlin noise, but belongs to a category called "[value noise](https://en.wikipedia.org/wiki/Value_noise#:~:text=Value%20noise%20is%20a%20type,Perlin%20noise%20and%20Simplex%20noise)." which has slightly different properties.

For p5.js’s purposes, I think this is good enough.

* It makes it clear that Perlin noise is the inspiration for `noise()`
* It makes it clear that `noise()` does not _strictly_ follow the definition of Perlin noise
* It clearly states that there are different properties, and offers a Wikipedia link for the curious reader to learn more about those differences if they care to

And all explained quickly, understandably, and without getting bogged down in details and making the documentation bloated. 👍

**Comment by GregStanton** [comment]
Great discussion everyone! Since @SableRaf and @davepagurek mentioned a potential noise mode [in Processing](https://github.com/processing/p5.js/issues/7430#issuecomment-2550585122) and [in p5](https://github.com/processing/p5.js/issues/7430#issuecomment-2548669744), I thought I'd share an API we came up with for spline modes. Adapting it for noise modes may provide significant benefits, and it may help resolve #6152 as well.
## Splines
For spline modes, the [currently proposed API](https://github.com/processing/p5.js/issues/6766#issuecomment-2560397417) is based on two functions: `splineMode(mode)` and `splineProperty(key, value)`. The latter functions as a getter when `value` is omitted.

Examples:

```
// Using default splines but adjusting tightness
splineProperty('tightness', 2);

// Using hobby curves with default params
splineMode(HOBBY);

// Using hobby curves with full config
splineMode(HOBBY);
splineProperty('tension', 1);
splineProperty('curlStart', 1);
splineProperty('curlEnd', 1);

// Using Yuskel curves and changing just the interpolator from the defaults
splineMode(YUSKEL);
splineProperty('interpolator', HYBRID);
```
One of the reasons for this proposal is that p5's `splineTightness()` function doesn't apply to all types of splines, so it may be an obstacle for add-on developers who want to implement new spline types.

Benefits:

* Provides an extensible foundation for add-on libraries or the core p5 library 
* Enables new features without a significant increase in the size of the overall p5 API
* Prevents conflicts with add-on libraries
* Leverages a simple syntax that already arises elsewhere in p5, and in native JavaScript

Some existing p5 features that use this syntax are listed at the end of this comment. This syntax is also present in the [`set()` method of `Map` objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/set) in native JavaScript.

## Noise
The same considerations that apply to splines also apply to noise. For example, [JNoise](https://github.com/Articdive/JNoise) is a Java library that supports a range of different noise types, [each with its own customizable features](https://github.com/Articdive/JNoise?tab=readme-ov-file#customizable-features). So, it might make sense to replace `noiseDetail()` with `noiseProperty()` in the same way that `splineTightness()` may be replaced with `splineProperty().`

## Other features
Here's a larger list of features that may benefit from a common pattern for configurable modes or general property setters/getters:

* [`splineProperty()`](https://github.com/processing/p5.js/issues/6766#issuecomment-2560397417) (proposed)
* [`setAttributes()`](https://p5js.org/reference/p5/setAttributes/)
* [`p5.XML.setAttribute()`](https://p5js.org/reference/p5.XML/setAttribute/)
* [`setUniform()`](https://p5js.org/reference/p5.Shader/setUniform/)
* [`vertexProperty()`](https://github.com/processing/p5.js/blob/9b48d049447378331ae977924a667e0d03bac2fd/src/shape/vertex.js#L1689) (2.0 feature)
* [`textProperty()`](https://github.com/processing/p5.js/pull/7356) (proposed 2.0 feature, implemented in dev-2.0; thanks to @dhowe for pointing this out to me)
* Specific features that might work better through general key-value setters:
  * `splineTightness()` (called [`curveTightness()`](https://p5js.org/reference/p5/curveTightness/) in p5.js 1.x)
  * [`noiseDetail()`](https://p5js.org/reference/p5/noiseDetail/)

To focus on just one example, the `setAttributes()` feature is a bit confusing, since it has a plural name, but `setAttributes(key, value)` only sets a single attribute. The name is also ambiguous, since it only sets attributes of the WebGL drawing context, but this isn't advertised in the name. Both of these issues could potentially be fixed by replacing `setAttributes()` with, say, `canvasProperty(key, value)`. Potentially, we could also introduce a `canvasMode()` function for consistency, but it may be preferable to continue setting the renderer via `createCanvas()`.

## Thoughts?

1. Any feedback on adding `noiseMode()` and `noiseProperty()` to match `splineMode()` and `splineProperty()`?
2. I can make a separate issue to help achieve consensus on a common pattern for customizable modes and general property setters across p5. I listed some relevant features above, but a full audit of the p5 API would take more time. Does anyone have anything else to add to the list?

**Comment by davepagurek** [comment]
I think `noiseDetail` could be applied universally, although different noises may want different default values. Whatever the base noise wave is, we turn it into fractal wave by summing multiple octaves at different scales by doing:

$$\sum_{i=0}^{n-1} \frac{f(x*k^i)}{k^i}$$

...where n is the number of octaves, and 1/k is the scale factor that p5 uses. The choice of nosie function f here could be anything, so any base noise would work.

Because of that, I don't think we need to remove `noiseDetail` as a general way of doing fractal noise. Adding methods to set other properties sounds good for when other noise modes need it though, e.g. for voronoi noise, switching from different distance metrics like Euclidean and Manhattan.

**Comment by GregStanton** [comment]
Thanks @davepagurek! Have you checked out [JNoise](https://github.com/Articdive/JNoise)? In addition to different noise types, each with their own customizable features, it has other features such as an octavation module:

>Octavation Module
>The underlying noise type to be octavated
> Amount of octaves
> Lacunarity
> Gain (Persistence)
> Fractal functions (FBM, Turbelent & Ridged)
>Seed incrementation per octave (Increases the seed by 1 each octave) 

I think gain/persistence is what p5 is calling "falloff." For the rest, I'm thinking that this much customization might be problematic with a `noiseDetail()` function, since it would potentially mean a long list of parameters. I haven't looked into any of this enough, though, and I know a breaking change involves trade-offs. So I'm not sure what's best.

**Comment by shiffman** [comment]
Hi everyone, thank you @nickmcintyre for bringing this thread to my attention! I'm teaching `noise()` in my Nature of Code class today and spent some time thinking through how I might revise my explanation in [Chapter 0 of the Nature of Code book](https://natureofcode.com/random/) to be as true to the implementation of `noise()` and its history. If anyone has any feedback you can follow the issue here in the NOC repo:  https://github.com/nature-of-code/noc-book-2/issues/1057.

Also, I would LOVE to help with anything related to a `noiseMode()` function in p5.js!

**Comment by quinton-ashley** [comment]
For the time being, p5's documentation of `noise` needs be changed, since it isn't Perlin Noise.

> Ken Perlin invented `noise()` while animating the original Tron film in the 1980s.

This also seems to imply that Kevin Perlin created p5's `noise` function or invented the concept of a noise function.

In q5.js I implemented `noiseMode` and made real Perlin Noise the default algorithm.

https://q5js.org/learn/#noise

I've referred to p5's noise algorithm (which is closer to value noise than perlin noise) as "blocky" noise, since that's how it looks when displayed in 2D slices at 1 octave.

**Comment by quinton-ashley** [comment]
For the time being, p5's documentation of `noise` needs be changed, since it isn't Perlin Noise.

> Ken Perlin invented `noise()` while animating the original Tron film in the 1980s.

This also seems to imply that Kevin Perlin created p5's `noise` function or invented the concept of a noise function.

If p5's noise belongs to a different category of noise then why allude to Perlin noise? Even saying it's similar to Perlin noise is incorrect. p5's noise algorithm is closer to value noise, so I refer to it as "blocky" noise, since that's how it looks when displayed in 2D slices at 1 octave. It only looks smooth at higher octaves.

In q5.js I implemented `noiseMode` and made real Perlin Noise the default algorithm.

https://q5js.org/learn/#noise

I noticed that p5/Processing's `noise()` is better at generating a good variety of noise, even when used with only 1 input parameter compared to Perlin noise though. However, since Perlin noise doesn't need octaves to look smooth, the performance cost is similar for higher quality results.

Simplex noise while marginally faster is significantly less visually appealing.
